### ISSUE TITLE
How to maintain session between API calls?

### ISSUE THREAD

**Issue by FlyingDiver** [bachya/simplisafe-python#350]
I've run into a roadblock trying to use simplipy in a more complex application.  In order to authenticate with SMS 2FA,  the user needs to do two UI actions - initiate the authentication process with credentials, then verify the SMS code.  This requires two round trips from the UI to the backend simplipy code.  I can't figure out how to maintain the ClientSession() between calls.  

If a do a new  `async with ClientSession() as session:` for each call, the `async_verify_2fa_sms` fails because it's expecting the original session from `async_from_credentials`.  Also, that call does not take `session` as a parameter.

Are there any examples that don't use console input() for authentication?

**Comment by bachya** [comment]
How about putting all your logic under that same `async with` statement?

```python
async with ClientSession() as session:
   # Do everything needed here...
```

**Comment by FlyingDiver** [comment]
Can't, this is a non-asyncio framework.  I can run all the asyncio code in it's own thread, if needed, but I still need to get UI data from the main framework.  So this code is long-running, and MAY need to re-authenticate.  And it definitely needs to authenticate on initial startup.  So the user needs to kick off the authentication, then input the SMS code in the UI and start the verification step.

**Comment by bachya** [comment]
> Can't, this is a non-asyncio framework.

That's a crucial bit of info. This library officially supports `asyncio`-powered use cases. Running `simplypy` in sync operates very differently – without knowing your application's architecture, I would anticipate regularly created new event loops to run `simplypy`; this, by design, going to require new `ClientSession` objects.

For instance, you mention `async_verify_2fa_sms` not taking a `session` parameter – that's by design, as you're not expected to be able to call that method unless you're within the `API` object (which already uses the same session everywhere).

**Comment by FlyingDiver** [comment]
Is it true that the API object returned from the auth methods, and used by everything else, is intimately tied to the ClientSession it was created in?  That it, if you want to use the API object, it always has to be in the ClientSession?

**Comment by bachya** [comment]
That's correct. More specifically, since the API object relies on `aiohttp`, it's bound to `asyncio` principles all the way down.

**Comment by FlyingDiver** [comment]
Ok, I'm going to have to see if I can figure out a new way to do this.

I'm relatively new to asyncio (and aiohttp).  Is there some technical reason the required ClientSession isn't a property of the API object?  Is there something about the "async with" paradigm that prevents that?  That's the concept that's new to me with this module.

**Comment by bachya** [comment]
> Is there some technical reason the required ClientSession isn't a property of the API object?

The `ClientSession` is, indeed, a property of the API object:

https://github.com/bachya/simplisafe-python/blob/b6648ce1a9bdd58ab8b74fa2a74d995f68fcc817/simplipy/api.py#L85

> Is there something about the "async with" paradigm that prevents that? That's the concept that's new to me with this module.

The `async with` syntax is an [asynchronous context manager](https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-3.html) and is a construct within Python itself (not `simplipy`).

---

Providing comprehensive education on `asyncio` and how it works is, unfortunately, beyond the scope of support I can provide here. Some materials that might be useful to you:

* [PEP 492 – Coroutines with async and await syntax](https://peps.python.org/pep-0492/)
* [Async IO in Python: A Complete Walkthrough](https://realpython.com/async-io-python/)
* [Asynchronous HTTP Requests in Python with aiohttp and asyncio](https://www.twilio.com/blog/asynchronous-http-requests-in-python-with-aiohttp)
* [Python Asyncio Part 5 – Mixing Synchronous and Asynchronous Code](https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-5.html)

In short, it's crucial to understand what an event loop is and how both sync and async code can (and can't) interact with it.

**Comment by FlyingDiver** [comment]
Thanks for the links, I'll review them.  I suspect this is some nuance of aiohttp that I'm unaware of.
